From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Fri, 22 Aug 2025 07:21:36 +0000
Subject: [PATCH 109/241] NVIDIA: SAUCE: tegra_hv: cert & misra fixes

Fixing below cert & misra errors:-
- CID 19420849: misra_c_2012_directive_4_14_violation: Passing tainted
  expression hvd->info to tegra_hv_cleanup, which uses it as a loop boundary.
- CID 19420337: misra_c_2012_directive_4_14_violation: Passing tainted
  expression qd->id to tegra_hv_add_ivc, which uses it as an offset.
- CID 27491737: cert_int30_c_violation: Unsigned integer operation
  ivc->num_frames * ivc->frame_size may wrap.
- CID 27491736: cert_int08_c_violation: Unsigned integer operation
  ivc->num_frames * ivc->frame_size may wrap.

Bug 5433066

Change-Id: I2ce11c8289ae8591de5918e6a3fd2ac79f31d76a
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3451318
Reviewed-by: svc-mobile-coverity <svc-mobile-coverity@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
Reviewed-by: Sandeep Trasi <strasi@nvidia.com>
(cherry picked from commit 38340e780b8762c102d0fc56bfa34ab58ab74dbf)

diff --git a/drivers/firmware/tegra/ivc.c b/drivers/firmware/tegra/ivc.c
index 75e50a7da4a0..bf2b45ad1980 100644
--- a/drivers/firmware/tegra/ivc.c
+++ b/drivers/firmware/tegra/ivc.c
@@ -645,6 +645,14 @@ static inline void *iosys_map_get_vaddr(const struct iosys_map *map)
 
 void tegra_ivc_clean_queue_data(struct tegra_ivc *ivc)
 {
+	size_t result;
+
+	/* Check for overflow in multiplication to satisfy CERT INT08-C */
+	if (check_mul_overflow(ivc->num_frames, ivc->frame_size, &result)) {
+		pr_err("%s: overflow in num_frames * frame_size calculation\n", __func__);
+		return;
+	}
+
 	iosys_map_memset(&ivc->rx.map, sizeof(struct tegra_ivc_header), 0, ivc->num_frames * ivc->frame_size);
 	iosys_map_memset(&ivc->tx.map, sizeof(struct tegra_ivc_header), 0, ivc->num_frames * ivc->frame_size);
 }
diff --git a/drivers/virt/tegra/tegra_hv.c b/drivers/virt/tegra/tegra_hv.c
index d62aaf826a1d..2f38cadb4a7d 100644
--- a/drivers/virt/tegra/tegra_hv.c
+++ b/drivers/virt/tegra/tegra_hv.c
@@ -781,6 +781,11 @@ static int tegra_hv_setup(struct tegra_hv_data *hvd)
 	for (i = 0; i < hvd->info->nr_queues; i++) {
 		const struct tegra_hv_queue_data *qd =
 				&ivc_info_queue_array(hvd->info)[i];
+		/* Validate qd->id before using as offset to satisfy MISRA C 2012 Directive 4.14 */
+		if (qd->id > hvd->max_qid) {
+			ERR("Invalid queue ID %u > max_qid %u\n", qd->id, hvd->max_qid);
+			return -EINVAL;
+		}
 		ret = tegra_hv_add_ivc(hvd, qd, i);
 		if (ret != 0) {
 			ERR("failed to add queue #%u\n", qd->id);
@@ -1210,7 +1215,10 @@ static int tegra_hv_probe(struct platform_device *pdev)
 
 	ret = tegra_hv_setup(hvd);
 	if (ret != 0) {
-		tegra_hv_cleanup(hvd);
+		/* Validate hvd->info before cleanup to satisfy MISRA C 2012 Directive 4.14 */
+		if (hvd->info && hvd->info->nr_areas <= MAX_NUM_GUESTS) {
+			tegra_hv_cleanup(hvd);
+		}
 		kfree(hvd);
 		return ret;
 	}
-- 
2.43.0

